### DISCUSSION TITLE
custom propagator needed ?

### DISCUSSION THREAD

**Discussion by gregy4** [google/or-tools#3303] [CP-SAT questions]
Hello,
  related to #2940 and approach with disjunctive constraints, I found number of constraints that guarantee minimal distance between intervals as main limiting factor when I reach million of constraints.

  Today I tried very naive propagator (not complete yet but I receive valid solutions for my data) in choco solver to just fail when distance between intervals is under minimal distance and to my surprise it works much better that cp-sat solution (seconds/minutes vs hours/days) on large instances. I use incremental propagator to know which variable was changed so I need to check only distances of newly assigned interval.

  Unfortunately custom propagators are not generally available in cp-sat solver. My idea is to copy simple propagator that just fails to verify that a propagator works as my naive propagator in choco solver. To be general I thought about alternative to nooverlap constraint with distance matrix as implemented in cp optimizer for example.

  Is there any other alternative that I'm not aware of ?  Of course I can try to minimize number of workers via assignment of workers before calculation of a schedule or limit number of possibilities in order of tasks but it seems to me as much harder to do.

Thanks,
Jan

---------------------------------------------------------------------------- 

```java
public class PropWalkingTime extends Propagator<IntVar>
{
    Task[] tasks;
    String[] msns;
    BoolVar[] ycut;

    public static void create(BoolVar Y[][], Task[] tasks, List<scheduler.task.Task> modelTasks, Solver solver)
    {
        int propCount = 0;
        for (int w = 0; w < Y[0].length; ++w) {
            List<BoolVar> ycut = new ArrayList<>();
            List<Task> ytasks = new ArrayList<>();
            List<String> msns = new ArrayList<>();
            for (int t = 0; t < Y.length; ++t) {
                if (Y[t][w] != null) {
                    ycut.add(Y[t][w]);
                    ytasks.add(tasks[t]);
                    msns.add(modelTasks.get(t).getMsn());
                }
            }
            if (!ycut.isEmpty()) {
                solver.post(new Constraint("", new PropWalkingTime(ycut.toArray(new BoolVar[0]), ytasks.toArray(new Task[0]),
                        msns.toArray(new String[0]))));
                ++propCount;
            }
        }

        System.out.println("wt propagators: " + propCount);
    }

    public PropWalkingTime(BoolVar[] ycut, Task[] tasks, String[] msns)
    {
        super(ycut, PropagatorPriority.VERY_SLOW, true);
        this.ycut = ycut;
        this.tasks = tasks;
        this.msns = msns;
    }

    @Override
    public void propagate(int changedIdx, int mask) throws ContradictionException
    {
        if (ycut[changedIdx].getLB() == 1) {
            Task changedTask = tasks[changedIdx];
            if (!changedTask.getStart().isInstantiated()) {
                return;
            }
            for (int t = 0; t < tasks.length; ++t) {
                if (changedIdx == t) continue;
                if (tasks[t].getStart().isInstantiated()
                        && ycut[t].getLB() == 1) {
                    if (msns[changedIdx].equals(msns[t])) continue;
                    Task task = tasks[t];
                    if (task.getStart().getValue() > changedTask.getStart().getValue()) {
                        if (task.getStart().getValue() - changedTask.getEnd().getValue() < getWalkingTime(changedIdx, t)) {
                            fails();
                        }
                    } else {
                        if (changedTask.getStart().getValue() - task.getEnd().getValue() < getWalkingTime(changedIdx, t)) {
                            fails();
                        }
                    }
                }
            }
        }
    }
```

**Comment by lperron** [comment]
and compared to the circuit model ?

**Reply by gregy4** [reply]
Circuit model was worst. It didn't work for more than 200 tasks. Disjunctive model worked ok until 500 tasks, with 1800 tasks I get to the point where used memory was close 100gb and a model contained around 11 millions disjunctive constraints. It was slow but I so some progress.

Of course when I have hundreds of circuits that relates to each other based on which worker is assigned to a task is hard.

With choco solver I received first valid solution in 6 minutes for 1500 tasks and used under 5gb of memory. Of course naive propagator is fragile to be lost during backtracking, probably reason why I don't have solution for 1800 tasks.

**Reply by lperron** [reply]
What to you call the disjunctive model ? one BoolVar with two implied precedence per pair of tasks ? or one no_overlap constraint per pair of tasks ?

**Reply by gregy4** [reply]
One BoolVar with two implied precedences per pair of optional intervals in case distance between tasks is not zero.
Optional interval represents possible assignment of a worker to a task, in case a worker is used task interval is equal to optional interval.
No_overlap constraint over all possible assignments of a worker to tasks guarantee that a worker is not used more than once.

**Comment by lperron** [comment]
If you have two optional intervals, you can still add the precedence
without any additional interval variable

```py
model.Add(interval_a_start >= interval_b_end +
delay_b_a).OnlyEnforceIf(presence_a, presence_b, precedence_a_b.Not())
model.Add(interval_b_start >= interval_a_end +
delay_a_b).OnlyEnforceIf(presence_a, presence_b, precedence_a_b)
```
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by gregy4** [reply]
Yes, that is true. I use it that way, only extra variable precedence_a_b for a pair is too much for larger instances so I use one variable for order of two task intervals instead two optional intervals.

To give some insight into my instances there's a short overview of one larger instance.

numworkers: 189
numtasks: 1815

tasks need 7 qualifications of workers, every task needs one or more workers with one qualification.
 qualification number_of_tasks
    q1              331
    q2              53
    q3              335
    q4              249
    q5              91
    q5              206

corresponding workers (every worker has only qualification)
   qualification number_of_workers
    q1                5
    q2               36
    q3               8
    q4               4
    q5               21
    q6              100
 
tasks are located in 8 places
   place       num_tasks
    place1     80
    place2    290
    place3    315
    place4     315
    place5    230
    place6    73
    place7    232
    place8   280

Every movement of worker between places is associated with fixed extra time.

189 (number of workers) no_overlap constraints is used to guarantee that a worker is not used more than one at same time.
Your code to guarantee extra time between tasks on different places is called 15825176 times, only in case pair of tasks is in different places.

There's a log of cp-sat solver, protofile is close to 2gb limit and I hit limits of available memory in my computer (128gb).


Starting CP-SAT solver v9.4.10641
Parameters: log_search_progress: true num_search_workers: 16

Initial optimization model '':
#Variables: 535789 (#ints:1 in objective)
  - 374046 in [0,1]
  - 161575 in [3555,4530]
  - 168 constants in {0,1,2,3,4,5,6,7,8,9,10,11,12 ... 284,287,292,324,332,342,388,401,427,480,492,568} 
#kBoolOr: 78972 (#literals: 157944)
#kCumulative: 113 (#intervals: 4542)
#kInterval: 80787 (#enforced: 78972)
#kLinMax: 1
#kLinear1: 166502 (#enforced: 162746)
#kLinear2: 31999559 (#enforced: 31997003)
#kLinear3: 80787 (#enforced: 78972)
#kLinearN: 3630 (#terms: 157944)
#kNoOverlap: 189 (#intervals: 78972, #optional: 78972)

Starting presolve at 21.40s
[ExtractEncodingFromLinear] #potential_supersets=3206 #potential_subsets=0 #at_most_one_encodings=0 #exactly_one_encodings=0 #unique_terms=0 #multiple_terms=0 #literals=0 time=0.44057s
[Probing] deterministic_time: 1.00001 (limit: 1) wall_time: 11.0003 (Aborted 103020/388415)
[Probing]  - new binary clause: 502707
[SAT presolve] num removable Booleans: 20054 / 535789
[SAT presolve] num trivial clauses: 0
[SAT presolve] [0s] clauses:597 literals:1194 vars:330 one_side_vars:330 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.00133025s] clauses:597 literals:1194 vars:330 one_side_vars:330 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.0191684s] clauses:597 literals:1194 vars:330 one_side_vars:330 simple_definition:0 singleton_clauses:0
[MaxClique] Merged 597(1194 literals) into 93(330 literals) at_most_ones.
[DetectDuplicateConstraints] #duplicates=2661 time=38.4323s
[DetectDominatedLinearConstraints] #relevant_constraints=2171 #work_done=30796 #num_inclusions=0 #num_redundant=0 time=0.437119s
[DetectOverlappingColumns] #processed_columns=0 #work_done=0 #nz_reduction=0 time=7.05978s
[ProcessSetPPC] #relevant_constraints=1908 #num_inclusions=0 work=231918 time=0.734326s
[Probing] deterministic_time: 1 (limit: 1) wall_time: 13.8093 (Aborted 292110/385836)
[Probing]  - new binary clause: 593938
[SAT presolve] num removable Booleans: 20054 / 535789
[SAT presolve] num trivial clauses: 0
[SAT presolve] [0s] clauses:57 literals:114 vars:114 one_side_vars:114 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.00125655s] clauses:57 literals:114 vars:114 one_side_vars:114 simple_definition:0 singleton_clauses:0
[SAT presolve] [0.0209484s] clauses:57 literals:114 vars:114 one_side_vars:114 simple_definition:0 singleton_clauses:0
[DetectDuplicateConstraints] #duplicates=0 time=36.6549s
[DetectDominatedLinearConstraints] #relevant_constraints=2171 #work_done=30796 #num_inclusions=0 #num_redundant=0 time=0.431526s
[DetectOverlappingColumns] #processed_columns=0 #work_done=0 #nz_reduction=0 time=6.83369s
[ProcessSetPPC] #relevant_constraints=1851 #num_inclusions=0 work=231468 time=0.762574s

Presolve summary:
  - 2238 affine relations were detected.
  - rule 'TODO linear2: contains a Boolean.' was applied 1194 times.
  - rule 'affine: new relation' was applied 2238 times.
  - rule 'at_most_one: transformed into max clique.' was applied 1 time.
  - rule 'bool_or: implications' was applied 79569 times.
  - rule 'bool_or: only one literal' was applied 14871 times.
  - rule 'cumulative: convert to no_overlap' was applied 2 times.
  - rule 'cumulative: divide demands and capacity by gcd' was applied 1 time.
  - rule 'cumulative: max profile is always under the min capacity' was applied 61 times.
  - rule 'cumulative: remove never conflicting intervals.' was applied 3 times.
  - rule 'cumulative: removed intervals with no demands' was applied 7 times.
  - rule 'cumulative: split into disjoint components' was applied 3 times.
  - rule 'deductions: 356645 stored' was applied 1 time.
  - rule 'dual: fix variable' was applied 189 times.
  - rule 'duplicate: removed constraint' was applied 2661 times.
  - rule 'enforcement: false literal' was applied 100321 times.
  - rule 'enforcement: true literal' was applied 21349 times.
  - rule 'lin_max: removed exprs' was applied 1 time.
  - rule 'lin_max: target domain reduced' was applied 1 time.
  - rule 'linear1: is boolean implication' was applied 78972 times.
  - rule 'linear1: without enforcement' was applied 3664 times.
  - rule 'linear2: implied ax + by = cte has only one solution' was applied 2 times.
  - rule 'linear: always true' was applied 25898 times.
  - rule 'linear: fixed or dup variables' was applied 82608 times.
  - rule 'linear: infeasible' was applied 14871 times.
  - rule 'linear: negative clause' was applied 597 times.
  - rule 'linear: positive equal one' was applied 3206 times.
  - rule 'linear: reduced variable domains' was applied 3665 times.
  - rule 'linear: remapped using affine relations' was applied 61671 times.
  - rule 'linear: simplified rhs' was applied 31928192 times.
  - rule 'presolve: 15228 unused variables removed.' was applied 1 time.
  - rule 'presolve: iteration' was applied 2 times.
  - rule 'variables: detect fully reified value encoding' was applied 13146 times.
  - rule 'variables: detect half reified value encoding' was applied 26381 times.
  - rule 'variables: merge equivalent var value encoding literals' was applied 391 times.
  - rule 'variables: removable enforcement literal' was applied 2588 times.

Presolved optimization model '':
#Variables: 515735 (#ints:1 in objective)
  - 1337 different domains in [0,4530] with a largest complexity of 3.
#kAtMostOne: 36 (#literals: 216)
#kBoolAnd: 57 (#enforced: 57) (#literals: 57)
#kCumulative: 53 (#intervals: 2610)
#kExactlyOne: 1603 (#literals: 72660)
#kInterval: 80787 (#enforced: 78972)
#kLinMax: 1
#kLinear1: 32029 (#enforced: 32029)
#kLinear2: 32061633 (#enforced: 32059674)
#kLinearN: 212 (#terms: 6312)
#kNoOverlap: 191 (#intervals: 78976, #optional: 78972)
